src/middlewared/middlewared/plugins/api_key.py

```diff
@@ -36,7 +45,14 @@ async def item_extend(self, item):
         Dict(
             "api_key_create",
             Str("name", required=True, empty=False),
-            register=True
+            List("allowlist", items=[
+                Dict(
+                    "allowlist_item",
+                    Str("method", required=True, enum=["GET", "POST", "PUT", "DELETE", "CALL", "SUBSCRIBE"]),
+                    Str("resource", required=True),
+                ),
+            ]),
+            register=True,
         )
```

Comment:
Won't this be O(N) lookups on each API request? What if we use dictionary `<resource>: {"GET": True, "POST": False, "PUT": False, "DELETE": True, "CALL": False, "SUBSCRIBE": True}`?

Though maybe ACL for this would be better defined through groups (separate db table) rather than a per-key basis (from a sysadmin standpoint). Then you could define permissions for a group and grant individual keys membership to that group.

Reply:
Having a separate table where API ACL is defined may be useful for situation where you decide to add additional authentication methods. It would also allow us to pre-populate with some sensible permissions / roles templates, like "backup operator".

Reply:
@anodos325 this would be nanoseconds

@william-gr do we want separate roles table now?

Reply:
A few things I dont quite understand.

We are only matching exact methods and not wildcards? It may be impractical to require a list of every single if not

Are we reading in the database for every call? That should probably be kept in memory for performance.

I agree with Andrew, we should have a forward thinking, the more we can do to make this future proof and better performance the better. We could use some of that for Kerberos auth in the future.

Also I am wondering how this pave the road for usage with the UI in the future. Someday we will want to restrict the UI using same backend.